x/evm/keeper/state_transition.go

```diff
@@ -232,16 +219,53 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
 		return nil, sdkerrors.Wrapf(err, "failed to refund gas leftover gas to sender %s", msg.From())
 	}
 
+	var bloomReceipt ethtypes.Bloom
 	if len(logs) > 0 {
 		res.Logs = types.NewLogsFromEth(logs)
 		// Update transient block bloom filter
 		bloom := k.GetBlockBloomTransient()
 		bloom.Or(bloom, big.NewInt(0).SetBytes(ethtypes.LogsBloom(logs)))
 		k.SetBlockBloomTransient(bloom)
+		bloomReceipt = ethtypes.BytesToBloom(bloom.Bytes())
 	}
 
 	k.IncreaseTxIndexTransient()
 
+	if !res.Failed() {
+		cumulativeGasUsed := res.GasUsed
+		if ctx.BlockGasMeter() != nil {
+			limit := ctx.BlockGasMeter().Limit()
+			consumed := ctx.BlockGasMeter().GasConsumed()
+			cumulativeGasUsed = uint64(math.Min(float64(cumulativeGasUsed+consumed), float64(limit)))
+		}
+
+		receipt := &ethtypes.Receipt{
+			Type:              tx.Type(),
+			PostState:         nil, // TODO: intermediate state root
+			Status:            ethtypes.ReceiptStatusSuccessful,
+			CumulativeGasUsed: cumulativeGasUsed,
+			Bloom:             bloomReceipt,
+			Logs:              logs,
+			TxHash:            txHash,
+			ContractAddress:   common.HexToAddress(res.ContractAddress),
+			GasUsed:           res.GasUsed,
+			BlockHash:         common.BytesToHash(ctx.HeaderHash()),
+			BlockNumber:       big.NewInt(ctx.BlockHeight()),
+			TransactionIndex:  uint(k.GetTxIndexTransient()),
+		}
+
+		// Only call hooks if tx executed successfully.
+		if err = k.PostTxProcessing(msg.From(), tx.To(), receipt); err != nil {
+			// If hooks return error, revert the whole tx.
+			res.VmError = types.ErrPostTxProcessing.Error()
+			k.Logger(ctx).Error("tx post processing failed", "error", err)
+		} else if commit != nil {
+			// PostTxProcessing is successful, commit the cache context
+			commit()
+			ctx.EventManager().EmitEvents(k.Ctx().EventManager().Events())
```

Comment:
hooks execution is moved under the `k.WithContext(ctx)`, this changed the context hooks executed in.
And `k.Ctx()` is same as `ctx` here, we'll lose events emitted by hooks, and re-emit events in `ctx.